pkg/operator/configobservation/encryption/key_controller_encryption.go

```diff
@@ -0,0 +1,243 @@
+		cond.Message = configError.Error()
+	}
+	if _, _, updateError := operatorv1helpers.UpdateStatus(c.operatorClient, operatorv1helpers.UpdateConditionFn(cond)); updateError != nil {
+		return updateError
+	}
+
+	return configError
+}
+
+func (c *EncryptionKeyController) handleEncryptionKey() error {
+	encryptionSecrets, err := c.secretLister.List(c.componentSelector)
+	if err != nil {
+		return err
+	}
+
+	encryptionState := getEncryptionState(encryptionSecrets, c.validGRs)
+
+	// make sure we look for all resources that we are managing
+	for gr := range c.validGRs {
+		if _, ok := encryptionState[gr]; !ok {
+			encryptionState[gr] = keysState{}
+		}
+	}
+
+	var errs []error
+	for gr, grKeys := range encryptionState {
+		keyID, ok := needsNewKey(grKeys)
+		if !ok {
+			continue
+		}
+
+		nextKeyID := keyID + 1
+		keySecret := c.generateKeySecret(gr, nextKeyID)
+		actualKeySecret, createErr := c.secretClient.Create(keySecret)
+		if errors.IsAlreadyExists(createErr) {
+			keyGR, _, keyID, ok := secretToKey(actualKeySecret, c.validGRs)
+			if valid := keyGR == gr && keyID == nextKeyID && ok; valid {
+				continue // we made this key earlier and our lister has not caught up
+			}
+			// delete the invalid secret so we do not get stuck
+			// TODO or we can just get stuck in degraded ?
+			errs = append(errs, c.secretClient.Delete(keySecret.Name, nil))
+		}
+		errs = append(errs, createErr)
+	}
+	// we do not filter using IsAlreadyExists as the remaining ones are actual errors
+	return utilerrors.FilterOut(utilerrors.NewAggregate(errs), errors.IsNotFound)
+}
+
+func (c *EncryptionKeyController) generateKeySecret(gr schema.GroupResource, keyID uint64) *corev1.Secret {
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("encryption-%s-%s-%s-%d", c.componentName, gr.Group, gr.Resource, keyID),
+			Namespace: operatorclient.GlobalMachineSpecifiedConfigNamespace,
+			Labels: map[string]string{
+				encryptionSecretComponent: c.componentName,
+
+				encryptionSecretGroup:    gr.Group,
+				encryptionSecretResource: gr.Resource,
+			},
+		},
+		Data: map[string][]byte{
+			encryptionSecretKeyData: newAES256Key(),
+		},
+	}
+}
+
+func needsNewKey(grKeys keysState) (uint64, bool) {
+	if len(grKeys.secretsMigratedNo) > 0 {
+		return 0, false
+	}
+
+	if len(grKeys.secrets) == 0 {
+		return 0, true
+	}
+
+	// TODO clean up logic to get this
+	lastMigrated := grKeys.secretsMigratedYes[len(grKeys.secretsMigratedYes)-1]
+	keyID, _ := secretToKeyID(lastMigrated) // TODO maybe store this
+
+	migrationTimestampStr := lastMigrated.Annotations[encryptionSecretMigratedTimestamp]
+	migrationTimestamp, err := time.Parse(time.RFC3339, migrationTimestampStr)
+	if err != nil {
+		return keyID, true // eh?
+	}
+
+	return keyID, time.Now().After(migrationTimestamp.Add(30 * time.Minute)) // TODO how often?
+}
+
+func newAES256Key() []byte {
+	b := make([]byte, 32) // AES-256 == 32 byte key
+	if _, err := rand.Read(b); err != nil {
```

Comment:
I'd handle the error, perhaps there are actually cases when this function might fail.

Reply:
It will never fail on linux, and the OAuth server has matched this behavior since the beginning of OpenShift.